pkg/context/context.go

```diff
@@ -78,3 +78,54 @@ func (c *Context) WithLogger(logger log.Logger) *Context {
 func (c *Context) L() log.Logger {
 	return c.Logger
 }
+
+type userCancelFlag struct{}
+
+// WithUserCancelFlag add an flag to context, to indicate if cancel() is called by user
+func WithUserCancelFlag(ctx context.Context) context.Context {
+	// no need for a lock, because write to this variable should happens before cancel() and read happens
+	// after ctx.Done()
+	cancelByUser := false
+	return context.WithValue(ctx, userCancelFlag{}, &cancelByUser)
```

Comment:
I think we'd better not introduce `context.WithValue`. It will make the code hard to understand 🤔 
WDYT @WangXiangUSTC

Reply:
WithValue but not exposed to caller

Reply:
I still think we needn't distinguish `user canceled` error and `error canceled` error. We can filter all the `context.Cancel` error (just log is Okay). The error that calls `context.Cancel` should be returned to user.

Reply:
reasonable, going to check whether there's severe context cancel

Reply:
addressed in https://github.com/pingcap/dm/pull/794/commits/7e1bea484ec08b3236a3ac0e82d5050ba19414ea